File: lua/bin/fetch_engine.lua
#!/usr/bin/env lua
-- fetch_engine — the one command a Lua dev runs to get the prebuilt engine.
--
--   lua bin/fetch_engine.lua [--tag vX.Y.Z] [--force] [--path]
--
-- Downloads + caches the prebuilt libselenium_core for this platform from the
-- project's GitHub releases (no Aether toolchain needed); the C extension then
-- loads it from the cache automatically. Mirrors Ruby's `rake selenium:fetch_engine`.

-- Resolve the module beside this script (bin/ -> ../src) without needing it installed.
local here = (arg[0] or ""):match("^(.*)/[^/]+$") or "."
package.path = here .. "/../src/?.lua;" .. package.path

local F = require("engine_fetcher")

local tag, force, path_only = F.ENGINE_VERSION, false, false
local i = 1
while arg[i] do
  local a = arg[i]
  if a == "--tag" then i = i + 1; tag = arg[i]
  elseif a:match("^%-%-tag=") then tag = a:sub(7)
  elseif a == "--force" then force = true
  elseif a == "--path" then path_only = true
  elseif a == "--help" or a == "-h" then
    io.write("usage: fetch_engine.lua [--tag vX.Y.Z] [--force] [--path]\n")
    os.exit(0)
  else
    io.stderr:write("fetch_engine: unknown argument: " .. a .. "\n"); os.exit(2)
  end
  i = i + 1
end

if path_only then
  print(F.cached_path(tag))
  os.exit(0)
end

io.stderr:write("selenium: fetching engine " .. F.asset_name(tag) .. " (" .. tag .. ") ...\n")
local ok, res = pcall(F.fetch, { tag = tag, force = force })
if not ok then
  io.stderr:write(tostring(res) .. "\n")
  os.exit(1)
end
io.stderr:write("selenium: engine ready at " .. res .. "\n")
io.stderr:write("selenium: require('selenium') will now load it automatically.\n")

File: dart/bin/fetch_engine.dart
/// `dart run selenium:fetch_engine` — the one command a Dart dev runs after
/// `dart pub get` to pull the prebuilt pure-Aether engine (libselenium_core)
/// for their platform from GitHub releases. No Aether toolchain, no compiler.
///
/// Flags:
///   --tag <vX.Y.Z>   pin a specific engine release (default: EngineFetcher.engineVersion)
///   --force          re-download even if a cached copy is present
///   --path           just print where the engine is (or would be) cached, no download
///   -h, --help       show this help
library;

import 'dart:io';

import 'package:selenium/src/engine_fetcher.dart';

Future<void> main(List<String> args) async {
  var tag = EngineFetcher.engineVersion;
  var force = false;
  var pathOnly = false;

  for (var i = 0; i < args.length; i++) {
    final a = args[i];
    switch (a) {
      case '--tag':
        if (i + 1 >= args.length) {
          stderr.writeln('selenium: --tag needs a value (e.g. --tag v0.8.0)');
          exit(2);
        }
        tag = args[++i];
        break;
      case '--force':
        force = true;
        break;
      case '--path':
        pathOnly = true;
        break;
      case '-h':
      case '--help':
        stdout.writeln(
          'Usage: dart run selenium:fetch_engine [--tag vX.Y.Z] [--force] [--path]\n'
          '  Downloads + caches the prebuilt engine (libselenium_core) for this\n'
          '  platform from GitHub releases. --path prints the cache location only.',
        );
        return;
      default:
        if (a.startsWith('--tag=')) {
          tag = a.substring('--tag='.length);
        } else {
          stderr.writeln('selenium: unknown argument: $a');
          exit(2);
        }
    }
  }

  if (pathOnly) {
    stdout.writeln(EngineFetcher.cachedPath(tag));
    return;
  }

  stderr.writeln(
      'selenium: fetching engine ${EngineFetcher.assetName(tag)} ($tag) …');
  try {
    final path = await EngineFetcher.fetch(tag: tag, force: force);
    stderr.writeln('selenium: engine ready at $path');
    stderr.writeln(
        "selenium: `import 'package:selenium/selenium.dart'` will now load it "
        'automatically.');
  } on FetchException catch (e) {
    stderr.writeln('selenium: engine fetch failed: ${e.message}');
    exit(1);
  }
}
